### DISCUSSION TITLE
1. 데이터베이스 - DBMS, 관계 데이터 모델

### DISCUSSION THREAD

**Discussion by zini9188** [Shortudy-10th/tech-for-developer#150] [질의 응답]
- 데이터베이스 시스템의 구성
- 관계 데이터 모델의 개념
- 무결성 제약조건
- 키
- 그 외 등등

**Comment by seongeun42** [comment]
## Q. 데이터베이스란 무엇인가요?

여러 응용 시스템들이 공용할 수 있도록 통합, 저장된 운영 데이터의 집합을 의미합니다.

**Reply by seongeun42** [reply]
## Q. 데이터베이스의 특성에 대해 말해주세요.

데이터베이스의 특성에는 수시로 오는 질의에 실시간으로 응답이 가능해야 하는 **실시간 접근성**, 항상 최신 상태를 유지하기 위한 **지속적인 변화**, 다수의 사용자가 동시에 같은 내용의 데이터를 이용할 수 있는 **동시 공용**, DB에 접근해 데이터를 참조할 때 주소값이 아닌 데이터의 내용으로 참조하는 **내용에 의한 참조**, 응용 프로그램에 영향을 주지 않고 DB의 논리적 구조를 변경할 수 있는 **논리적 독립성** 등이 있습니다.

**Reply by LireEruel** [reply]
## Q. 데이터베이스는 왜 사용하나요?

데이터베이스 존재 이전에는 파일 시스템을 이용하여 데이터를 관리하였습니다. 이 방식은 파일에 접근하는 방식이 응용 프로그램 내에 표현되므로 응용 프로그램과 데이터 간의 의존관계가 존재하게 되어 데이터의 구조가 변경되면 기존 프로그램과 데이터를 함꼐 변경해야 했습니다.
따라서 파일 단위로 저장할 때, 데이터 종속성  문제와 중복성, 데이터 무결성 문제등이 존재하기 때문에 데이터베이스로 관리하기 시작했습니다.

**Reply by LireEruel** [reply]
## Q. 파일 시스템과 데이터베이스(DB)는 뭐가 다른건가요?
 파일 시스템은 데이터를 파일과 디렉토리로 구성되며, 데이터의 구조화와 관련이 적습니다. 데이터 간의 관계를 나타내기도 어려우며 데이터 간의 중복이 발생할 수 있으며 일관성도 유지하기 어렵습니다.
 DB는 데이터를 테이블, 레코드 및 열의 형태로 구조화하여 저장합니다. 이러한 구조화된 데이터는 데이터베이스 스키마를 따라 정의되며, 데이터의 일관성과 유효성을 유지하기 쉽게 만듭니다.

**Comment by zini9188** [comment]
### Q. RDMBS를 사용하며 나타나는 단점을 설명하고 이를 극복하기 위한 방안을 말해주세요.
시스템이 커짐에 따라 테이블간의 관계가 많아지고 데이터를 가져오기 위한 JOIN 연산이 많아져 복잡한 쿼리들이 생성되게 됩니다.  이러한 이유로 스키마의 변경에 유연하지 않습니다. 

이를 극복하기 위해 NoSQL을 사용할 수 있습니다. NoSQL은 스키마가 없기 때문에 유연한 데이터 구조를 가지고 있습니다. 또한 JOIN 연산이 없기 때문에 쿼리가 단순해지고 성능이 개선되며 읽기 및 쓰기 연산에서 좋은 성능을 기대할 수 있습니다.

**Reply by zini9188** [reply]
### Q. 그렇다면 NoSQL을 사용하는 것이 좋은건가요?

NoSQL의 경우 여러 장점을 가지고 있지만 여러 요인에 따라 RDBMS를 사용하는 것이 더 좋은 상황이 있습니다.

복잡한 관계를 가지는 데이터의 경우나 트랜잭션이 필요한 경우, 수정이 많이 일어나는 데이터의 경우 RDBMS를 선택하는 것이 좋습니다. 하지만 유연한 데이터 구조를 가지고 있거나, 대규모 분산 환경 및 대량의 읽기/쓰기 작업을 처리하는 환경에서는 NoSQL을 선택하는 것이 좋습니다.

**Comment by seongeun42** [comment]
## Q. DBMS란 무엇인가요?

사용자와 데이터의 중재자로서 모든 사용자들이 데이터베이스를 공용할 수 있도록 관리해주는 소프트웨어 시스템을 말합니다. 데이터베이스의 구성, 접근 방법, 관리 등에 대한 모든 책임을 DBMS가 가집니다.

**Comment by watty-exclamationMark** [comment]
## Q. 데이터베이스의 키란 무엇인가요?
데이터베이스에서 키란, 각각의 레코드 (데이터 항목)을 고유하게 식별하거나, 데이터 간의 관계를 설정하는 데 사용되는 값이나 속성(column) 을 말합니다.

**Reply by watty-exclamationMark** [reply]
## Q. 데이터베이스의 키에서 기본키와 후보키의 차이는 무엇인가요?
기본키는 후보키에서 선택된 하나의 키로, 하나의 테이블에 여러 후보키가 있을 수 있어도, 기본키는 하나만 존재해야 합니다.

**Comment by watty-exclamationMark** [comment]
## Q. 데이터베이스 무결성이란 무엇인가요?
데이터베이스 내의 데이터가 정확하고 일관성을 유지하며, 데이터베이스의 정의에 따라 제약 조건을 준수하는 상태를 말합니다.

**Comment by Jihn0118** [comment]
## Q. 데이터베이스 무결성 제약 조건 3가지만 말씀해 주세요.
개체 무결성, 참조 무결성, 도메인 무결성 제약 조건이 있습니다.

**Reply by Jihn0118** [reply]
## Q. 각각에 대한 설명
1. 개체 무결성(Entity Integrity)
**기본키 제약 조건**을 통해 유지됩니다. 
릴레이션에서 **기본키를 구성하는 속성은 NULL 값이나 중복 값을 가질 수 없습니다**.

2. 참조 무결성(Referential Integrity)
**외래키**를 사용하여 유지됩니다. 
외래 키 값은 관련 테이블의 주 키 값을 참조해야 하며, 이를 통해 **릴레이션은 참조할 수 없는 외래키 값을 가질 수 없습니다**.

3. 도메인 무결성(Domain Integrity)
각 컬럼의 데이터 유형과 제약 조건을 정의하여 해당 열의 데이터가 **특정한 도메인 내에 있음을 보장**합니다.
예) 성별 속성의 도메인은 '남'과 '여'로 그 외의 값은 입력하지 못합니다.

**Comment by sikutisa** [comment]
## Q. 데이터베이스 엔진은 무엇인가요? 
데이터베이스 엔진(Database Engine) 또는 스토리지 엔진(Storage Engine)은 데이터베이스 관리 시스템(DBMS)이 데이터베이스에 대해 데이터를 삽입, 추출, 업데이트 및 삭제하는데 사용하는 기본 소프트웨어 컴포넌트입니다.

**Comment by seongeun42** [comment]
## Q. 데이터 독립성에 대해 설명해주세요.

**데이터 독립성**이란 데이터의 논리적 구조나 물리적 구조가 변경되더라도 응용 프로그램은 영향을 받지 않는 것을 말합니다. 데이터 독립성에는 논리적 데이터 독립성과 물리적 데이터 독립성이 있습니다.

**Reply by seongeun42** [reply]
## Q. 논리적 독립성과 물리적 독립성에 대해 설명해주세요.

**논리적 독립성**이란 응용 프로그램에 영향을 주지 않고 데이터베이스의 논리적 구조를 변경할 수 있는 특성을 의미합니다. **물리적 독립성**이란 응용 프로그램이나 데이터베이스의 논리적 구조에 영향을 주지 않고 데이터베이스의 물리적 구조를 변경할 수 있는 특성을 의미합니다. 위의 데이터 독립성은 각 구조 간의 **사상**을 통해 가능합니다.

**Reply by seongeun42** [reply]
## 사상에 대해 설명해주세요.

**사상**이란 데이터 독립성을 제공하기 위해 각 구조 간의 대응 관계에 대해 정의한 것을 말합니다. 사상에는 외부-개념 사상과 개념-내부 사상이 있습니다.

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/e619ce18-bc16-4c89-b5fc-818089487b86" width=500>
